src/pkg/backup/details/testdata/in_deets.go

```diff
@@ -0,0 +1,366 @@
+// of that location.
+func (id *InDeets) RemoveLocation(set, locationRef string) {
+	id.getSet(set).RemoveLocation(locationRef)
+}
+
+// MoveLocation takes the LAST elemet in the fromLocation (and all)
+// children matching the prefix, and relocates it as a child of toLocation.
+// ex: MoveLocation("/a/b/c", "/d") will move all entries with the prefix
+// "/a/b/c" into "/d/c".  This also deletes all "/a/b/c" entries and children.
+// assumes item IDs don't change across the migration.  If item IDs do change,
+// that difference will need to be handled manually by the caller.
+// returns the base folder's new location (ex: /d/c)
+func (id *InDeets) MoveLocation(set, fromLocation, toLocation string) string {
+	return id.getSet(set).MoveLocation(fromLocation, toLocation)
+}
+
+func (id *InDeets) RenameLocation(set, fromLocation, toLocation string) {
+	id.getSet(set).RenameLocation(fromLocation, toLocation)
+}
+
+// Subset produces a new locSet containing only Items and Locations
+// whose location matches the locationPfx
+func (id *InDeets) Subset(set, locationPfx string) *locSet {
+	return id.getSet(set).Subset(locationPfx)
+}
+
+// ---------------------------------------------------------------------------
+// whatSet helpers for extracing a set identifier from an arbitrary repoRef
+// ---------------------------------------------------------------------------
+
+type whatSet func(details.Entry) (string, error)
+
+// common whatSet parser that extracts the service category from
+// a repoRef.
+func CategoryFromRepoRef(ent details.Entry) (string, error) {
+	p, err := path.FromDataLayerPath(ent.RepoRef, false)
+	if err != nil {
+		return "", err
+	}
+
+	return p.Category().String(), nil
+}
+
+// common whatSet parser that extracts the driveID from a repoRef.
+func DriveIDFromRepoRef(ent details.Entry) (string, error) {
+	p, err := path.FromDataLayerPath(ent.RepoRef, false)
+	if err != nil {
+		return "", err
+	}
+
+	odp, err := path.ToOneDrivePath(p)
+	if err != nil {
+		return "", err
+	}
+
+	return odp.DriveID, nil
+}
+
+// ---------------------------------------------------------------------------
+// helpers and comparators
+// ---------------------------------------------------------------------------
+
+func CheckBackupDetails(
+	t *testing.T,
+	ctx context.Context, //revive:disable:context-as-argument
+	backupID model.StableID,
+	ws whatSet,
+	ms *kopia.ModelStore,
+	ssr streamstore.Reader,
+	expect *InDeets,
+) {
+	deets, result := GetDeetsInBackup(t, ctx, backupID, "", "", path.UnknownService, ws, ms, ssr)
+
+	t.Log("details entries in result")
+
+	for _, ent := range deets.Entries {
+		if ent.Folder == nil {
+			t.Log(ent.LocationRef)
+			t.Log(ent.ItemRef)
+		}
+
+		assert.Truef(
+			t,
+			strings.HasPrefix(ent.RepoRef, expect.RRPrefix),
+			"all details should begin with the expected prefix\nwant: %s\ngot:  %s",
+			expect.RRPrefix, ent.RepoRef)
+	}
+
+	for set := range expect.Sets {
+		assert.Subsetf(
```

Comment:
For our tests do we want a subset or a something like `ElementsMatch`? With a subset it seems like it won't error out on folders that should have been deleted (removed from the backup) but weren't? Would we need the deleted set if we could do an exact match?

If I remember right, we're backing up a specific directory so we shouldn't have to worry about other tests messing with state we'll see right? Do we intend to use this for stuff where we can't carefully control the data set?

Reply:
Yeah, waffled on this a lot.  I stuck with Subset because I want this to be generically usable by any test that needs to verify details, and not just these tests, without concern over external cross-pollination.  I think I'll add a flag that lets the caller opt for exact-match, since test cases like this should provide some concrete expectations about scope.